platforms/chibios/drivers/vendor/RP/RP2040/ps2_vendor.c

```diff
@@ -0,0 +1,277 @@
+        osalSysUnlockFromISR();
+    }
+
+    if (irqs & (PIO_IRQ0_INTF_SM0_TXNFULL_BITS << state_machine)) {
+        pio_set_irq0_source_enabled(pio, pis_sm0_tx_fifo_not_full + state_machine, false);
+        osalSysLockFromISR();
+        osalThreadResumeI(&tx_thread, MSG_OK);
+        osalSysUnlockFromISR();
+    }
+}
+
+void ps2_host_init(void) {
+    ibqObjectInit(&pio_rx_queue, false, pio_rx_buffer, sizeof(uint32_t), BUFFER_SIZE, NULL, NULL);
+    uint pio_idx = pio_get_index(pio);
+
+    hal_lld_peripheral_unreset(pio_idx == 0 ? RESETS_ALLREG_PIO0 : RESETS_ALLREG_PIO1);
+
+    state_machine = pio_claim_unused_sm(pio, true);
+    if (state_machine < 0) {
+        dprintln("ERROR: Failed to acquire state machine for PS/2!");
+        ps2_error = PS2_ERR_NODATA;
+        return;
+    }
+
+    uint offset = pio_add_program(pio, &ps2_program);
+
+    pio_sm_config c = pio_get_default_sm_config();
+    sm_config_set_wrap(&c, offset + PS2_WRAP_TARGET, offset + PS2_WRAP);
+
+    // Set pindirs to input (output enable is inverted below)
+    pio_sm_set_consecutive_pindirs(pio, state_machine, PS2_DATA_PIN, 2, true);
+    sm_config_set_clkdiv(&c, (float)clock_get_hz(clk_sys) / (200.0f * KHZ));
+    sm_config_set_set_pins(&c, PS2_DATA_PIN, 2);
+    sm_config_set_out_pins(&c, PS2_DATA_PIN, 1);
+    sm_config_set_out_shift(&c, true, true, 10);
+    sm_config_set_in_shift(&c, true, true, 11);
+    sm_config_set_jmp_pin(&c, PS2_CLOCK_PIN);
+    sm_config_set_in_pins(&c, PS2_DATA_PIN);
+
+    // clang-format off
+    iomode_t pin_mode = PAL_RP_PAD_IE |
+                        PAL_RP_GPIO_OE |
+                        PAL_RP_PAD_SLEWFAST |
+                        PAL_RP_PAD_DRIVE12 |
+                        // Invert output enable so that pindirs=1 means input
+                        // and indirs=0 means output. This way, out pindirs
+                        // works correctly with the open-drain PS/2 interface.
+                        // Setting pindirs=1 effectively pulls the line high,
+                        // due to the pull-up resistor, while pindirs=0 pulls
+                        // the line low.
+                        PAL_RP_IOCTRL_OEOVER_DRVINVPERI |
+                        (pio_idx == 0 ? PAL_MODE_ALTERNATE_PIO0 : PAL_MODE_ALTERNATE_PIO1);
+    // clang-format on
+
+    palSetLineMode(PS2_DATA_PIN, pin_mode);
+    palSetLineMode(PS2_CLOCK_PIN, pin_mode);
+
+    pio_set_irq0_source_enabled(pio, pis_sm0_rx_fifo_not_empty + state_machine, true);
+    pio_sm_init(pio, state_machine, offset, &c);
+
+#if defined(PS2_PIO_USE_PIO1)
+    nvicEnableVector(RP_PIO1_IRQ_0_NUMBER, CORTEX_MAX_KERNEL_PRIORITY);
+#else
+    nvicEnableVector(RP_PIO0_IRQ_0_NUMBER, CORTEX_MAX_KERNEL_PRIORITY);
+#endif
+
+    pio_sm_set_enabled(pio, state_machine, true);
+}
+
+static int bit_parity(int x) {
+    x ^= x >> 16;
+    x ^= x >> 8;
+    x ^= x >> 4;
+    x ^= x >> 2;
+    x ^= x >> 1;
+    return (~x) & 1;
+}
```

Comment:
The function can be replaced with calls to `!__builtin_parity()` notice that odd parity is 1 and even parity is 0. Thus the checks need an inversion.

Reply:
Oh, didn't knew about this builtin function. Thanks!

Reply:
Great!